### PR TITLE
Added chains compliant kaniko task

### PR DESCRIPTION
**Pull request by pxp928**
# Changes

Added chains compliant Kaniko task. Addresses issue #908 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Follows the [authoring recommendations](https://github.com/tektoncd/catalog/blob/main/recommendations.md)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality of task changed or new task added)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Complies with [Catalog Organization TEP][TEP], see [example]. **Note** [An issue has been filed to automate this validation][validation]
  - [x] File path follows  `<kind>/<name>/<version>/name.yaml`
  - [x] Has `README.md` at `<kind>/<name>/<version>/README.md`
  - [x] Has mandatory `metadata.labels` - `app.kubernetes.io/version` the same as the `<version>` of the resource
  - [x] Has mandatory `metadata.annotations` `tekton.dev/pipelines.minVersion`
  - [x] mandatory `spec.description` follows the convention

          ```

          spec:
            description: >-
              one line summary of the resource

              Paragraph(s) to describe the resource.
          ```

_See [the contribution guide](https://github.com/tektoncd/catalog/blob/master/CONTRIBUTING.md)
for more details._

---

[TEP]: https://github.com/tektoncd/community/blob/master/teps/0003-tekton-catalog-organization.md
[example]: https://github.com/tektoncd/catalog/tree/master/task/git-clone/0.1
[validation]:  https://github.com/tektoncd/catalog/issues/413

### CODE REVIEW

File: task/kaniko/0.6/kaniko.yaml [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    - --context=$(workspaces.source.path)/$(params.CONTEXT)  # The user does not need to care the workspace and the source.
+    - --destination=$(params.IMAGE)
+    - --digest-file=$(results.IMAGE_DIGEST.path)
+    # kaniko assumes it is running as root, which means this example fails on platforms
+    # that default to run containers as random uid (like OpenShift). Adding this securityContext
+    # makes it explicit that it needs to run as root.
+    securityContext:
+      runAsUser: 0
+  - name: write-url
+    image: bash
```
Comment (vinamra28) [inline task/kaniko/0.6/kaniko.yaml:67]: can you please use fully qualified image name ? eg `gcr.io/kaniko-project/executor:v1.5.1@sha256:c6166717f7fe0b7da44908c986137ecfeab21f31ec3992f6e128fff8a94be8a5`

**Review by vinamra28 — CHANGES_REQUESTED** [review]
/ok-to-test

File: task/kaniko/0.6/README.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+> - [Kaniko website](https://github.com/GoogleCloudPlatform/kaniko)
+
+kaniko is meant to be run as an image, `gcr.io/kaniko-project/executor:v1.5.1`. This
+makes it a perfect tool to be part of Tekton. This task can also be used with Tekton Chains to
+attest and sign the image.
+
+## Changelog
+
+- Use `--digest-file` flag to create the `Results` with the digest of a built image
+
```
[the remark this answers is not in the document]
Comment (pxp928) [reply task/kaniko/0.6/README.md:19]: Sorry about that. Updated the changelog.

File: task/kaniko/0.6/README.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+## Install the Task
+
+```
+kubectl apply -f https://raw.githubusercontent.com/tektoncd/catalog/main/task/kaniko/0.5/kaniko.yaml
```
[the remark this answers is not in the document]
Comment (pxp928) [reply task/kaniko/0.6/README.md:23]: Changed.

File: task/kaniko/0.6/kaniko.yaml [hunk 4]
```diff
@@ -0,0 +1,72 @@
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: kaniko
+  labels:
+    app.kubernetes.io/version: "0.5"
```
[the remark this answers is not in the document]
Comment (pxp928) [reply task/kaniko/0.6/kaniko.yaml:6]: changed.

File: task/kaniko/0.6/kaniko.yaml [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    description: Digest of the image just built.
+  - name: IMAGE_URL
+    description: URL of the image just built.
+  steps:
+  - name: add-dockerfile
+    workingDir: $(workspaces.source.path)
+    image: bash
+    script: |
+      set -e
+      echo "FROM alpine@sha256:69e70a79f2d41ab5d637de98c1e0b055206ba40a8145e7bddb55ccc04e13cf8f" | tee $(params.DOCKERFILE)
```
[the remark this answers is not in the document]
Comment (pxp928) [reply task/kaniko/0.6/kaniko.yaml:51]: Removed as not necessary. It was included in Chains kaniko task but tested without and it works.

File: task/kaniko/0.6/kaniko.yaml [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    # makes it explicit that it needs to run as root.
+    securityContext:
+      runAsUser: 0
+  - name: write-url
+    image: bash
+    script: |
+      set -e
+      echo $(params.IMAGE) | tee $(results.IMAGE_URL.path)
+    securityContext:
+      runAsUser: 0
```
[the remark this answers is not in the document]
Comment (pxp928) [reply task/kaniko/0.6/kaniko.yaml:72]: removed root user as not needed.

File: task/kaniko/0.6/kaniko.yaml (hunk 1 above)
Comment (pxp928) [reply task/kaniko/0.6/kaniko.yaml:67]: added fully qualified image for bash.

**Comment by pxp928** [comment]
is the build test just failing because of the lint errors?

**Comment by vdemeester** [comment]
> is the build test just failing because of the lint errors?

```
task/kaniko/0.6/kaniko.yaml
  65:1      error    too many blank lines (1 > 0)  (empty-lines)
```

Seems like yamllint isn't happy about too many empty-lines 😉

**Comment by pxp928** [comment]
the shellchecker might have to be updated to ignore this?
```
echo "$(params.IMAGE)" | tee "$(results.IMAGE_URL.path)"
```
it fails on and is needed for tekton to expand the parameter

**Comment by pxp928** [comment]
Is there a way we can approve this PR? Its been sitting in queue for a long time.

**Comment by vinamra28** [comment]
hi @pxp928, apologies for being late on this 🙏 I was hoping that someone else would also review this but I'll merge it
/lgtm

**Pull request merged**